### ISSUE TITLE
Feature Request: Dynamic User-Defined Custom Python Blocks

### ISSUE THREAD

**Issue by gouravroy2110** [srinjoy-2005/ArchIDE#24]
## Summary
Currently, all neural network blocks are hardcoded in `backend/blocks/`. Users cannot introduce novel PyTorch modules, custom loss functions, or experimental layers without modifying ArchIDE's internal source code.

## Proposed Solution
- **Custom Blocks Directory**: Users place custom block Python files in `custom_blocks_dir` (as specified in `arch.config.json`).
- **Dynamic Registration**: The backend scans and dynamically imports custom blocks using `importlib`, registering them into `_BLOCK_INSTANCES` and serving them through `/api/blocks`.
- **Blocks Palette Integration**: The frontend renders custom blocks in the right-hand Blocks palette under a "Custom" category with auto-inferred ports, parameters, and color-coding.
- **Error Isolation**: Syntax errors or broken imports in custom block files must be caught and reported gracefully without crashing the IDE backend.

**Comment by gouravroy2110** [comment]
## Developer Implementation Guide: Dynamic User-Defined Custom Python Blocks

### 1. Overview
This feature allows users to write custom PyTorch blocks in Python files located in their workspace (`custom_blocks_dir` from `arch.config.json`), and have them automatically discovered, validated, and rendered in the ArchIDE Blocks palette.

### 2. Custom Block Definition Example

Users create a Python file, e.g. `workspace/custom_blocks/swiglu.py`:

```python
import torch
import torch.nn as nn
import torch.nn.functional as F
from backend.blocks.base import BaseBlock, PortDef, ParamDef

class SwiGLUBlock(BaseBlock):
    id = "custom_swiglu"
    name = "SwiGLU"
    category = "Custom"
    color = "#10b981"
    is_functional = True

    inputs = [
        PortDef(id="in", name="Input", var_hint="x")
    ]
    outputs = [
        PortDef(id="out", name="Output", var_hint="out")
    ]
    params = [
        ParamDef(
            name="dim",
            type="int",
            default=-1,
            description="Dimension along which to split tensor"
        )
    ]

    def infer_shapes(self, input_shapes, params):
        in_shape = input_shapes.get("in", ("ANY",))
        if in_shape == ("ANY",):
            return {"out": ("ANY",)}
        dim = params.get("dim", -1)
        out_shape = list(in_shape)
        if isinstance(out_shape[dim], int):
            out_shape[dim] = out_shape[dim] // 2
        return {"out": tuple(out_shape)}

    def emit_init(self, layer_name, params):
        return ""

    def emit_forward(self, layer_name, input_vars, output_vars, params):
        x = input_vars.get("in", "x")
        dim = params.get("dim", -1)
        out = output_vars.get("out", "out")
        return f"x1, x2 = {x}.chunk(2, dim={dim})\n        {out} = F.silu(x1) * x2"
```

### 3. Backend Discovery & Registration Architecture

1. **Loader Module (`backend/custom_blocks_loader.py`)**:
   - Reads `custom_blocks_dir` from `arch.config.json` (or default `./custom_blocks`).
   - Scans for all `*.py` files in that directory.
   - Uses `importlib.util.spec_from_file_location` to dynamically import each module.
   - Inspects module classes: if `issubclass(cls, BaseBlock)` and `cls is not BaseBlock`, instantiates and registers into `_BLOCK_INSTANCES`.
2. **Error Isolation & Diagnostics**:
   - If a custom block has a syntax error or failed import:
     - Catch exception and store error in `CUSTOM_BLOCK_ERRORS: dict[str, str]`.
     - DO NOT crash the server.
     - Expose errors via `/api/blocks/diagnostics` so the frontend can show a warning badge in the palette.
3. **Endpoint Updates**:
   - `GET /api/blocks`: Returns built-in blocks + successfully loaded custom blocks.
   - `POST /api/blocks/reload`: Triggers hot-reload of custom blocks on file change.

### 4. Frontend Integration

1. **Blocks Palette (`src/components/BlocksPalette.tsx`)**:
   - Automatically groups custom blocks under a distinct `"Custom"` category with a special badge.
   - Displays a warning banner if any custom block files failed to compile.
2. **CustomNode Component**:
   - Renders custom block nodes using the block's `color`, `inputs`, and `outputs` defined in the schema.

### 5. Implementation Tasks
- [ ] Implement `backend/custom_blocks_loader.py` with `load_custom_blocks(custom_dir)`.
- [ ] Hook `load_custom_blocks()` into FastAPI startup in `backend/main.py`.
- [ ] Add `POST /api/blocks/reload` endpoint for hot-reloading custom blocks without restarting server.
- [ ] Add tests in `backend/tests/test_custom_blocks_loader.py` validating registration, shape inference, and syntax error isolation.
